forem/forem
# Using a Form Object that will persist for the Creator Settings Form

## What type of PR is this? (check all applicable)


## Description

Demo video can be found [here](https://www.loom.com/share/1e9100f4dd714c3892f00aabac8dffc9).

The problem that we had on this page and on many other pages in the Admin, is that once a user submits their data in a form, and if it errors out, it shows a flash of an error and then a blank form, forcing users to start over. 

In this scenario, on the Creator Settings Page, we can actually save the transactions that have passed as they are not dependant on each other as we previously thought. Hence, when we submit a form, and something fails, we want to show an error message stating what has failed and also display the information that the user has filled in that has already been saved to the database.

In addition, we have set default values on `Settings` class that we should show as defaults on the form. 

After chatting to Michale Kohl, we we decided to use Form Objects for this -  it cleans up a lot of aspects moves conditional callbacks and validations out of models so we don’t have to worry about accidentally triggering them from unrelated actions. In addition, it provides with a good way to update the multiple Settings at the same time.

You can read more about Form Object [here](https://nimblehq.co/blog/lets-play-design-patterns-form-objects)

## Related Tickets & Documents
Closes https://github.com/forem/forem/issues/15651

## QA Instructions, Screenshots, Recordings

For this PR its best to drop your db, migrate and enable your feature flag to test as a new user with the default Settings Applied.

```
rails db:drop RAILS_ENV=development
rails db:create RAILS_ENV=development
rails db:migrate RAILS_ENV=development

FeatureFlag.enable(:creator_onboarding)
```

Now, run your server and navigate to http://localhost:3000/ and complete the "Create Account" form. After completing the "Create Account" form, you should be brought to the new Creator Setup page, ${baseUrl}/admin/creator_settings/new?referrer=${baseUrl}.

You can play around with the form and make sure that the ffg values are saved as per you update. FIll out your values on the form and then chcek them in the rails console: 

```
Settings::Community.community_name
Settings::UserExperience.primary_brand_color_hex
Settings::UserExperience.public
Settings::Authentication.invite_only_mode
Settings::General.original_logo
Settings::General.resized_logo
```

To mimic an error, and then test the other values get saves besides the logo you can save an image thats dimensions are greater than 4096px x 4096px as per the demo. 
 

### UI accessibility concerns?

None, there were no frontend changes.

## Added/updated tests?

      have not been included_

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

      [Storybook](https://storybook.forem.com/) (for Crayons components)
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
      or in a [forem.dev](http://forem.dev) post
- [x] This change does not need to be communicated now and will be communicated when the feature is released.

## [optional] Are there any post deployment tasks we need to perform?

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](https://media.giphy.com/media/YRtLgsajXrz1FNJ6oy/giphy-downsized-large.gif)
